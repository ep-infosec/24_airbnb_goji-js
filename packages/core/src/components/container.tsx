import React, { createContext, PropsWithChildren, useContext } from 'react';
import { Container } from '../container';

export const ContainerContext = createContext<Container | undefined>(undefined);

export const ContainerProvider = ({
  children,
  container,
}: PropsWithChildren<{ container: Container }>) => {
  const { Provider } = ContainerContext;

  return <Provider value={container}>{children}</Provider>;
};

export const useContainer = () => {
  const container = useContext(ContainerContext);
  if (!container) {
    throw new Error('Cannot found `containerContext`. This might be an internal error in GojiJS.');
  }

  return container;
};
