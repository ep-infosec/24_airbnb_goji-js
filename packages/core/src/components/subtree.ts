import React, { createElement, Fragment, CSSProperties } from 'react';
import { TYPE_SUBTREE } from '../constants';

export const useSubtree = process.env.GOJI_TARGET === 'wechat' || process.env.GOJI_TARGET === 'qq';

export const subtreeMaxDepthFromConfig = (process.env.GOJI_MAX_DEPTH as any as number) ?? 10;

export const Subtree = ({
  unsafe_className: className,
  unsafe_style: style,
  ...restProps
}: React.PropsWithChildren<{ unsafe_className?: string; unsafe_style?: CSSProperties }>) => {
  if (useSubtree) {
    return createElement(TYPE_SUBTREE, {
      className,
      style,
      ...restProps,
    });
  }
  return createElement(Fragment, restProps);
};
