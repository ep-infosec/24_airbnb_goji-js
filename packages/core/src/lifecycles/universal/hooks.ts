import { useContext, useRef } from 'react';
import { OnLoadOptions } from '../types';
import { UniversalHooksContext, UniversalHooksContextType } from './context';
import { useImmediatelyEffect } from '../../utils/effects';
import { EventChannel } from '../../utils/eventChannel';

const createUniversalLifecycleHook =
  <T = undefined>(
    channelSelector: (context: UniversalHooksContextType) => EventChannel<any, any>,
  ) =>
  (callback: (data: T) => void, deps?: Array<any>) => {
    const universalHooksContext = useContext(UniversalHooksContext);
    if (!universalHooksContext) {
      throw new Error(
        'Cannot found `universalHooksContext`. This might be an internal error in GojiJS.',
      );
    }

    const callbackRef = useRef<(data: T) => void>(callback);
    // only update callbackRef when deps changed
    useImmediatelyEffect(() => {
      callbackRef.current = callback;
    }, deps);

    useImmediatelyEffect(
      () => channelSelector(universalHooksContext).on(data => callbackRef.current(data)),
      [],
    );
  };

export const useLoadOptions = createUniversalLifecycleHook<OnLoadOptions>(
  _ => _.universalLifecycleChannel.loadOptions,
);
export const useVisibility = createUniversalLifecycleHook<boolean>(
  _ => _.universalLifecycleChannel.visibility,
);
