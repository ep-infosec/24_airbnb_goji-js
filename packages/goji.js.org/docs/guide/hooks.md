---
sidebar_position: 3
---

# Hooks

## Hooks

Goji support all React build-in hooks like `useState`, `useEffect` and `useContext`. You can find
these hooks [here](https://reactjs.org/docs/hooks-reference.html).

Goji also support hooks from third-part library. This is an example for `react-redux`:

```tsx
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { myDataStringSelector } from './path/to/myDataStringSelector';

const Comp = () => {
  const dispatch = useDispatch();
  const dataString = useSelector(myDataStringSelector);
  useEffect(() => {
    dispatch({
      type: 'APP_DID_MOUNT',
    });
  }, []);

  return <View>{dataString}</View>;
};
```

In the next section, you will see Goji build-in [lifecycle hooks](./lifecycle.md).
