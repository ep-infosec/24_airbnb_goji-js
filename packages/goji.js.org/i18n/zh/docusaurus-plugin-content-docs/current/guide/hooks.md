---
sidebar_position: 3
---

# 钩子

## Hooks

Goji 支持所有React内置的钩子，例如 `useState`, `useEffect` 和 `useContext`。 你可以在[这里](https://reactjs.org/docs/hooks-reference.html)找到所有钩子。

Goji还支持第三方库中的钩子。 这是使用 `react-redux` 的例子：

```tsx
import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { myDataStringSelector } from './path/to/myDataStringSelector';

const Comp = () => {
  const dispatch = useDispatch();
  const dataString = useSelector(myDataStringSelector);
  useEffect(() => {
    dispatch({
      type: 'APP_DID_MOUNT',
    });
  }, []);

  return <View>{dataString}</View>;
};
```

在下一节中，你将看到Goji内置的 [生命周期钩子](./lifecycle.md)。
