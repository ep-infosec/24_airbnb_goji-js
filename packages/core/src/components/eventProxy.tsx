import {
  OnResizeOptions,
  OnScrollOptions,
  OnShareAppMessageOptions,
  OnShareAppMessageReturn,
  OnTabItemTapOptions,
} from '../lifecycles/types';
import { EventChannel } from '../utils/eventChannel';
import { useContainer } from './container';

export interface EventProxy {
  lifecycleChannel: {
    onLoad: EventChannel<Record<string, string>, void>;
    onReady: EventChannel;
    onShow: EventChannel;
    onHide: EventChannel;
    onUnload: EventChannel;
    onPullDownRefresh: EventChannel;
    onReachBottom: EventChannel;
    onPageScroll: EventChannel<OnScrollOptions, void>;
    onShareAppMessage: EventChannel<OnShareAppMessageOptions, OnShareAppMessageReturn>;
    onResize: EventChannel<OnResizeOptions, void>;
    onTabItemTap: EventChannel<OnTabItemTapOptions, void>;
    onTitleClick: EventChannel;
    onOptionMenuClick: EventChannel;
    onPopMenuClick: EventChannel;
    onPullIntercept: EventChannel;
  };
  internalChannels: {
    rendered: EventChannel<number>;
    exportedComponentUpdate: EventChannel<any>;
  };
}

export const createEventProxy = (): EventProxy => ({
  lifecycleChannel: {
    onLoad: new EventChannel<Record<string, string>, void>(),
    onReady: new EventChannel(),
    onShow: new EventChannel(),
    onHide: new EventChannel(),
    onUnload: new EventChannel(),
    onPullDownRefresh: new EventChannel(),
    onReachBottom: new EventChannel(),
    onPageScroll: new EventChannel<OnScrollOptions, void>(),
    onShareAppMessage: new EventChannel<OnShareAppMessageOptions, OnShareAppMessageReturn>(),
    onResize: new EventChannel<OnResizeOptions, void>(),
    onTabItemTap: new EventChannel<OnTabItemTapOptions, void>(),
    onTitleClick: new EventChannel(),
    onOptionMenuClick: new EventChannel(),
    onPopMenuClick: new EventChannel(),
    onPullIntercept: new EventChannel(),
  },
  internalChannels: {
    rendered: new EventChannel<number>(),
    exportedComponentUpdate: new EventChannel<any>(),
  },
});

export const useEventProxy = () => {
  const container = useContainer();

  return container.eventProxy;
};
