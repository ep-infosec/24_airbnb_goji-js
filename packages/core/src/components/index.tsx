import React, { PropsWithChildren } from 'react';
import { UniversalHooksProvider } from '../lifecycles/universal/provider';
import { PortalProvider } from '../portal';
import { Container } from '../container';
import { ContainerProvider } from './container';
import { PartialProvider } from '../partial';

export const GojiProvider = ({
  children,
  container,
}: PropsWithChildren<{
  container: Container;
}>) => (
  <ContainerProvider container={container}>
    <UniversalHooksProvider>
      <PortalProvider>
        <PartialProvider>{children}</PartialProvider>
      </PortalProvider>
    </UniversalHooksProvider>
  </ContainerProvider>
);
