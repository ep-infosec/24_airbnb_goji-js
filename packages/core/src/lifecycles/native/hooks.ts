import {
  OnScrollOptions,
  OnShareAppMessageOptions,
  OnResizeOptions,
  OnTabItemTapOptions,
} from '../types';
import { useImmediatelyEffect } from '../../utils/effects';
import { EventProxy, useEventProxy } from '../../components/eventProxy';
import { EventChannel } from '../../utils/eventChannel';

const createLifecycleHook =
  <T = undefined>(channelSelector: (eventProxy: EventProxy) => EventChannel<any, any>) =>
  (callback: (data: T) => void, deps?: Array<any>) => {
    const eventProxyContext = useEventProxy();
    // must use `useImmediatelyEffect` instead of `useEffect`
    useImmediatelyEffect(() => {
      if (!eventProxyContext) {
        throw new Error(
          'Cannot found `eventProxyContext`. This might be an internal error in GojiJS.',
        );
      }
      return channelSelector(eventProxyContext).on(callback as any);
    }, deps); // eslint-disable-line react-hooks/exhaustive-deps
  };

export const useOnReady = createLifecycleHook(_ => _.lifecycleChannel.onReady);
export const useOnUnload = createLifecycleHook(_ => _.lifecycleChannel.onUnload);
export const useOnPullDownRefresh = createLifecycleHook(_ => _.lifecycleChannel.onPullDownRefresh);
export const useOnReachBottom = createLifecycleHook(_ => _.lifecycleChannel.onReachBottom);
export const useOnPageScroll = createLifecycleHook<OnScrollOptions>(
  _ => _.lifecycleChannel.onPageScroll,
);
export const useOnShareAppMessage = createLifecycleHook<OnShareAppMessageOptions>(
  _ => _.lifecycleChannel.onShareAppMessage,
);
export const useOnResize = createLifecycleHook<OnResizeOptions>(_ => _.lifecycleChannel.onResize);
export const useOnTabItemTap = createLifecycleHook<OnTabItemTapOptions>(
  _ => _.lifecycleChannel.onTabItemTap,
);
// alipay only
export const useOnTitleClick = createLifecycleHook(_ => _.lifecycleChannel.onTitleClick);
export const useOnOptionMenuClick = createLifecycleHook(_ => _.lifecycleChannel.onOptionMenuClick);
export const useOnPopMenuClick = createLifecycleHook(_ => _.lifecycleChannel.onPopMenuClick);
export const useOnPullIntercept = createLifecycleHook(_ => _.lifecycleChannel.onPullIntercept);

// FIXME: internal hooks, consider to move to other folder
export const useInternalComponentUpdate = createLifecycleHook<object>(
  _ => _.internalChannels.exportedComponentUpdate,
);
