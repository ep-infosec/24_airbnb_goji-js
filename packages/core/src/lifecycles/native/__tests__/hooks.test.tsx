import React, { useState } from 'react';
import { act, create } from 'react-test-renderer';
import { View } from '../../..';
import { useOnPageScroll } from '../hooks';
import { GojiProvider } from '../../../components';
import { createEventProxy } from '../../../components/eventProxy';

describe('native lifecycles', () => {
  it('onLoad works', () => {
    let onPageScrollOptions: any;
    const App = () => {
      const [value, setValue] = useState(0);
      useOnPageScroll(options => {
        onPageScrollOptions = options;
        setValue(options.scrollTop);
      });
      return <View>value is {value}</View>;
    };
    const eventProxy = createEventProxy();
    const wrapper = create(
      <GojiProvider container={{ eventProxy } as any}>
        <View className="text">hello, world!</View>
        <App />
      </GojiProvider>,
    );

    expect(wrapper.toJSON()).toMatchSnapshot();

    // trigger onLoad
    const mockOnPageScroll = { scrollTop: 1024 };
    act(() => {
      eventProxy.lifecycleChannel.onPageScroll.emit(mockOnPageScroll);
    });

    expect(onPageScrollOptions).toEqual(mockOnPageScroll);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
