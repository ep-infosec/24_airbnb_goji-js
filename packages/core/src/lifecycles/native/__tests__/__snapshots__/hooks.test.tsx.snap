// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`native lifecycles onLoad works 1`] = `
Array [
  <view
    className="text"
  >
    hello, world!
  </view>,
  <view>
    value is 
    0
  </view>,
]
`;

exports[`native lifecycles onLoad works 2`] = `
Array [
  <view
    className="text"
  >
    hello, world!
  </view>,
  <view>
    value is 
    1024
  </view>,
]
`;
